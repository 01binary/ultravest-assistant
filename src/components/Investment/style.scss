.investment {
	border: 1px solid gray;
}
