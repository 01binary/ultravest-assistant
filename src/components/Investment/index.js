import { h } from 'preact';
import classNames from 'obj-str';
import Units from '../Units';
import getMixWeights from '../../selectors/getMixWeights';
import withInvestmentOrQuery from '../../enhancers/withInvestmentOrQuery';
import timelineStyle from '../App/style/timeline';
import formStyle from '../App/style/forms';
import calcStyle from '../App/style/calc';
import style from './style';

/**
 * Investment parameters
 * @param {Object} flask - The flask props provided by withFlask.
 * @param {string} preset - The investment preset provided by withInvestment.
 * @param {Object[]} presets - The investment presets provided by withInvestment.
 * @param {function} handleQueryPresetChange - The handler provided by withInvestment.
 * @returns {JSX.Element} - A React stateless component.
 */
export const Investment = ({
	preset,
	presets,
	flask,
	handleQueryPresetChange }) => (
	<article class={classNames({
		[timelineStyle.timeline]: true,
		[style.investment]: true
	})}
	>
		<h2>investment</h2>

		<section class={formStyle.group}>
			<section class={formStyle.control}>
				<select
					id="investment-preset"
					name="investment-preset"
					onChange={handleQueryPresetChange}
				>
					{ Object.keys(presets).map(presetOption => (
						<option selected={presetOption === preset}>
							{presetOption}
						</option>
					))}

					{
						console.log('got preset in Investment comp', preset)
					}
				</select>
				<label for="investment-preset">
					ratio
				</label>
			</section>
		</section>

		<output class={classNames({
			[formStyle.group]: true,
			[calcStyle.calc]: true
		})}
		>
			{ getMixWeights(flask, preset, presets).map(mix => (
				<dl class={formStyle.control}>
					<dt>{mix.component}</dt>
					<dd>
						{mix.grams}
						<Units alt="grams">g</Units>
					</dd>
				</dl>
			))}
		</output>
	</article>
);

export default withInvestmentOrQuery(Investment);
