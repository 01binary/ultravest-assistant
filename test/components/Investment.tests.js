import { h } from 'preact';
import { deep } from 'preact-render-spy';
import Investment from '../../src/components/Investment';

describe('Investment', () => {

	let wrapper, props;

	beforeAll(() => {
		props = {
			investment: {
				preset: expectedName,
				presets: {
					[expectedName]: expected
				}
			},
			flask: {
				preset: 'another preset',
				presets: {
					'another preset': {
						diameter: 4,
						height: 6
					}
				},
				diameter: 4,
				height: 6
			},
			handleInvestmentPresetChange: jest.fn()
		};

		wrapper = deep(
			<Investment {...props} />
		);
	});

	afterAll(() => {
		wrapper = null;
	});

	test('should render', () => {
		expect(wrapper).toMatchSnapshot();
	});

	test('should bind preset value', () => {
		const first = wrapper.find('option')[0];
		expect(first.children[0]).toBe(expectedName);
		expect(first.attributes.selected).toBe(true);
	});

	test('should bind change preset handler', () => {
		wrapper.find('#investment-preset')[0].attributes.onChange();
		expect(props.handleInvestmentPresetChange.mock.calls.length)
			.toBe(1);
	});

	const expectedName = 'test preset';
	const expected = {
		investment: 2.1,
		water: 3.5
	};
});
