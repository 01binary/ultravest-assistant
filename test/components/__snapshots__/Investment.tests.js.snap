// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Investment should render 1`] = `
preact-render-spy (1 nodes)
-------
<article class="investment">
  <h2>investment</h2>
  <label for="investment-ratio">
    ratio
    <select
      name="investment-ratio"
      class="investmentRatio"
      value="test preset"
      onChange={[Function mockConstructor]}
    >
      <option selected={true}>test preset</option>
    </select>
  </label>
  <output>
    <dl>
      <dt>investment</dt>
      <dd>158</dd>
    </dl>
    <dl>
      <dt>water</dt>
      <dd>264</dd>
    </dl>
  </output>
</article>

`;
